### PR TITLE
 refactor(lyd): refactor dt_policy in new pipeline and add img input support

### PR DESCRIPTION
**Pull request by AltmanD**
Refactor DT to new pipeline.
Add img input support for atari.

### CODE REVIEW

File: ding/policy/dt.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+        self.act_dim = self._cfg.model.act_dim
+
+        self._learn_model = self._model
+        self._optimizer = torch.optim.AdamW(self._learn_model.parameters(), lr=lr, weight_decay=wt_decay)
+
+        self._scheduler = torch.optim.lr_scheduler.LambdaLR(
+            self._optimizer, lambda steps: min((steps + 1) / warmup_steps, 1)
+        )
+
+        self.max_env_score = -1.0
```
Comment (PaParaZz1) [inline ding/policy/dt.py:120]: remove this

File: ding/policy/dt.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+            if self._cfg.model.continuous:
+                action_target = action_target.view(-1, self.act_dim)[traj_mask > 0]
+                action_loss = F.mse_loss(action_preds, action_target)
+            else:
+                action_target = action_target.view(-1)[traj_mask > 0]
+                action_loss = F.cross_entropy(action_preds, action_target)
+
+        self._optimizer.zero_grad()
+        action_loss.backward()
+        torch.nn.utils.clip_grad_norm_(self._learn_model.parameters(), 0.25)
```
Comment (PaParaZz1) [inline ding/policy/dt.py:173]: add this argument (0.25) into config

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/d4rl/config/__init__.py [hunk 3]
```diff
@@ -1,3 +1,3 @@
-from .hopper_cql_config import hopper_cql_config
-from .hopper_expert_cql_config import hopper_expert_cql_config
-from .hopper_medium_cql_config import hopper_medium_cql_config
+# from .hopper_cql_config import hopper_cql_config
+# from .hopper_expert_cql_config import hopper_expert_cql_config
+# from .hopper_medium_cql_config import hopper_medium_cql_config
```
[the remark this answers is not in the document]
Comment (AltmanD) [reply dizoo/d4rl/config/__init__.py:3]: configs these want import are not exist

File: ding/policy/dt.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        # (bool) Whether the RL algorithm is on-policy or off-policy.
+        on_policy=False,
+        # (bool) Whether use priority(priority sample, IS weight, update priority)
+        priority=False,
+        # (int) N-step reward for target q_value estimation
+        obs_shape=4,
+        action_shape=2,
+        # encoder_hidden_size_list=[128, 128, 64],
+        dataset='medium',  # medium / medium-replay / medium-expert
+        rtg_scale=1000,  # normalize returns to go
```
Comment (PaParaZz1) [inline ding/policy/dt.py:48]: add comments for each parameters in default config

File: ding/policy/dt.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+        self.rtg_scale = self._cfg.rtg_scale  # normalize returns to go
+        self.rtg_target = self._cfg.rtg_target  # max target reward_to_go
+        self.state_dim = self._cfg.model.state_dim
+        self.act_dim = self._cfg.model.act_dim
+        self.eval_batch_size = self._cfg.evaluator_env_num
+        self.max_eval_ep_len = self._cfg.max_eval_ep_len
+        self.context_len = self._cfg.context_len  # K in decision transformer
+        
+        self.t = [0 for _ in range(self.eval_batch_size)]
+        if not self._cfg.model.continuous:
```
Comment (PaParaZz1) [inline ding/policy/dt.py:200]: reverse if-else branch

File: ding/policy/dt.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+                    self.actions[i] = torch.zeros((self.max_eval_ep_len, self.act_dim), dtype=torch.float32, device=self.device)
+                if self.cfg.env_type == 'atari':
+                    self.states[i] = torch.zeros((self.max_eval_ep_len,) + tuple(self.state_dim), dtype=torch.float32, device=self.device)
+                    self.running_rtg[i] = self.rtg_target
+                else:
+                    self.states[i] = torch.zeros((self.max_eval_ep_len, self.state_dim), dtype=torch.float32, device=self.device)
+                    self.running_rtg[i] = self.rtg_target / self.rtg_scale
+                self.rewards_to_go[i] = torch.zeros((self.max_eval_ep_len, 1), dtype=torch.float32, device=self.device)
+
+    def get_d4rl_normalized_score(self, score, env_name):
```
Comment (PaParaZz1) [inline ding/policy/dt.py:319]: where do we use this method

File: ding/policy/dt.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+        self.t = [0 for _ in range(self.eval_batch_size)]
+        if not self._cfg.model.continuous:
+            self.actions = torch.zeros((self.eval_batch_size, self.max_eval_ep_len, 1), dtype=torch.long, device=self.device)
+        else:
+            self.actions = torch.zeros((self.eval_batch_size, self.max_eval_ep_len, self.act_dim), dtype=torch.float32, device=self.device)
+        if self.cfg.env_type == 'atari':
+            self.states = torch.zeros((self.eval_batch_size, self.max_eval_ep_len,) + tuple(self.state_dim), dtype=torch.float32, device=self.device)
+            self.running_rtg = [self.rtg_target for _ in range(self.eval_batch_size)]
+        else:
+            self.running_rtg = [self.rtg_target / self.rtg_scale for _ in range(self.eval_batch_size)]
```
Comment (PaParaZz1) [inline ding/policy/dt.py:208]: remove `rtg_scale` argument

File: ding/policy/dt.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+                        timesteps[i] = self.timesteps[i, self.t[i] - self.context_len + 1:self.t[i] + 1]
+                    states[i] = self.states[i, self.t[i] - self.context_len + 1:self.t[i] + 1]
+                    actions[i] = self.actions[i, self.t[i] - self.context_len + 1:self.t[i] + 1]
+                    rewards_to_go[i] = self.rewards_to_go[i, self.t[i] - self.context_len + 1:self.t[i] + 1]
+            # if not self._cfg.model.continuous:
+            #     actions = one_hot(actions.squeeze(-1), num=self.act_dim)
+            _, act_preds, _ = self._eval_model.forward(timesteps, states, actions, rewards_to_go)
+            del timesteps, states, actions, rewards_to_go
+            
+            if self.cfg.env_type == 'atari':
```
Comment (PaParaZz1) [inline ding/policy/dt.py:263]: polish this part

File: ding/policy/dt.py [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+        dataset='medium',  # medium / medium-replay / medium-expert
+        rtg_scale=1000,  # normalize returns to go
+        max_eval_ep_len=1000,  # max len of one episode
+        batch_size=64,  # training batch size
+        wt_decay=1e-4,
+        warmup_steps=10000,
+        max_train_iters=200,
+        context_len=20,
+        log_dir='DI-engine/dizoo/box2d/lunarlander/dt_log_1000eps',   
+        learn=dict(
```
Comment (PaParaZz1) [inline ding/policy/dt.py:56]: remove sub level

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/utils/data/dataset.py [hunk 10]
```diff
@@ -328,48 +328,258 @@ class D4RLTrajectoryDataset(Dataset):
...
-        # normalize states
-        for traj in self.trajectories:
-            traj['observations'] = (traj['observations'] - self.state_mean) / self.state_std
+    def __init__(self, cfg: dict) -> None:
+        dataset_path = cfg.policy.collect.data_path
+        rtg_scale = cfg.policy.rtg_scale
+        self.context_len = cfg.policy.context_len
+        self.env_type = cfg.policy.env_type
+
+        if 'hdf5' in dataset_path:
```
Comment (PaParaZz1) [inline ding/utils/data/dataset.py:337]: add env type comment

File: ding/utils/data/dataset.py [hunk 11]
```diff
@@ -328,48 +328,258 @@ class D4RLTrajectoryDataset(Dataset):
...
+            start_index = 0
+            rtg = np.zeros_like(stepwise_returns)
+            for i in done_idxs:
+                i = int(i)
+                curr_traj_returns = stepwise_returns[start_index:i]
+                for j in range(i-1, start_index-1, -1): # start from i-1
+                    rtg_j = curr_traj_returns[j-start_index:i-start_index]
+                    rtg[j] = sum(rtg_j)
+                start_index = i
+            print('max rtg is %d' % max(rtg))
```
Comment (PaParaZz1) [inline ding/utils/data/dataset.py:563]: use logging

File: ding/utils/data/dataset.py [hunk 12]
```diff
@@ -378,56 +588,134 @@ def get_d4rl_dataset_stats(self, env_d4rl_name: str) -> Dict[str, list]:
...
+            states = states / 255.
+            actions = torch.tensor(self.actions[idx:done_idx], dtype=torch.long).unsqueeze(1) # (block_size, 1)
+            rtgs = torch.tensor(self.rtgs[idx:done_idx], dtype=torch.float32).unsqueeze(1)
+            timesteps = torch.tensor(self.timesteps[idx:idx+1], dtype=torch.int64).unsqueeze(1)
             traj_mask = torch.ones(self.context_len, dtype=torch.long)
-
+            return timesteps, states, actions, rtgs, traj_mask
+        
+
+class FixedReplayBuffer(object):
```
Comment (PaParaZz1) [inline ding/utils/data/dataset.py:660]: can we merge this class into the above class?

File: ding/utils/data/dataset.py [hunk 13]
```diff
@@ -378,56 +588,134 @@ def get_d4rl_dataset_stats(self, env_d4rl_name: str) -> Dict[str, list]:
...
+    if replay_buffer is not None:
+        self._replay_buffers = [replay_buffer]
+        self.add_count = replay_buffer.add_count
+        self._num_replay_buffers = 1
+        self._loaded_buffers = True
+
+  def _load_buffer(self, suffix):
+    """Loads a OutOfGraphReplayBuffer replay buffer."""
+    try:
+        from dopamine.replay_memory import circular_replay_buffer
```
Comment (PaParaZz1) [inline ding/utils/data/dataset.py:698]: can we remove requirement to `dopamine`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/atari/entry/atari_dt_main.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+from dizoo.atari.envs import AtariEnv
+from dizoo.atari.config.pong_dt_config import main_config, create_config
+
+
+def main():
+    # If you don't have offline data, you need to prepare if first and set the data_path in config
+    # For demostration, we also can train a RL policy (e.g. SAC) and collect some data
+    logging.getLogger().setLevel(logging.INFO)
+    cfg = compile_config(main_config, create_cfg=create_config, auto=True)
+    # ding_init(cfg)
```
[the remark this answers is not in the document]
Comment (AltmanD) [reply dizoo/atari/entry/atari_dt_main.py:24]: using 'DistributedWriter' cause different experiment log into one directory when use w&b

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**